framework/src/media/HttpStream.cpp

```diff
@@ -0,0 +1,184 @@
+/******************************************************************
+ *
+ * Copyright 2018 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************/
+
+#include <curl/curl.h>
+#include <curl/easy.h>
+#include <debug.h>
+
+#include "HttpStream.h"
+
+namespace media {
+namespace stream {
+
+#define SET_OPTION(curl, opt, val) \
+	do {\
+		auto result = curl_easy_setopt(curl, opt, val);\
+		if (result != CURLE_OK) {\
+			meddbg("curl_easy_setopt(%d...) failed!\n", (int)opt);\
+			return false;\
+		}\
+	} while (0)
+
+int HttpStream::mInitialzeCount = 0;
+
+std::shared_ptr<HttpStream> HttpStream::create()
+{
+	std::shared_ptr<HttpStream> stream(new HttpStream());
+	if (!stream->init()) {
+		meddbg("HttpStream instance create failed!\n");
+		return nullptr;
+	}
+
+	return stream;
+}
+
+HttpStream::HttpStream() :
+	mCurl(nullptr), mHttpHeaders(nullptr)
+{
+}
+
+HttpStream::~HttpStream()
+{
+	cleanup();
+}
+
+bool HttpStream::addHeader(const std::string &header) {
+	mHttpHeaders = curl_slist_append(mHttpHeaders, header.c_str());
+	if (!mHttpHeaders) {
+		meddbg("curl_slist_append failed, header %s\n", header.c_str());
+		return false;
+	}
+
+	return true;
+}
+
+bool HttpStream::setHeaderCallback(CallbackFunc callback, void *userdata)
+{
+	SET_OPTION(mCurl, CURLOPT_HEADERFUNCTION, callback);
+	SET_OPTION(mCurl, CURLOPT_HEADERDATA, userdata);
+	return true;
+}
+
+bool HttpStream::setWriteCallback(CallbackFunc callback, void *userdata)
+{
+	SET_OPTION(mCurl, CURLOPT_WRITEFUNCTION, callback);
+	SET_OPTION(mCurl, CURLOPT_WRITEDATA, userdata);
```

Comment:
Do not you also need settings of CURLOPT_BUFFERSIZE?
A value smaller than the buffer size should be set.

Reply:
this `userdata` is not a buffer address, and we write data to a stream buffer, I think we do not need a CURLOPT_BUFFERSIZE setting.

Reply:
same as above.